_posts/2020-09-30-sparklyr-1.4.0-released/sparklyr-1.4.0.Rmd

```diff
@@ -0,0 +1,368 @@
+```
+
+- Copy all 105 values into a Spark dataframe named `sdf`
+
+```{r echo=TRUE, eval=FALSE}
+library(sparklyr)
+
+sc <- spark_connect(master = "local", version = "3.0.0")
+sdf <- copy_to(sc, data.frame(value = c(sample_values, outliers)))
+```
+
+- We can then apply `ft_robust_scaler()` to obtain the standardized value for each input:
+
+```{r echo=TRUE, eval=FALSE}
+scaled <- sdf %>%
+  ft_vector_assembler("value", "input") %>%
+  ft_robust_scaler("input", "scaled") %>%
+  dplyr::pull(scaled) %>%
+  unlist()
+```
+
+- Plotting the result shows the non-outlier data points are scaled to values that still more or less form a bell-shaped distribution, as expected, so the scaling is robust against influences of the outliers:
+
+```{r echo=TRUE, eval=FALSE}
+library(ggplot2)
+
+ggplot(data.frame(scaled = scaled), aes(x = scaled)) + geom_histogram(binwidth = 0.2)
+```
+
+![](images/scaled.png){#id .class width=40% height=40%}
+
+- Finally, we can compare the distribution of the scaled values above with the distribution of z-scores of all input values, and notice how scaling the input with only mean and standard deviation would have caused noticeable skewness which the robust scaler has successfully avoided:
+
+
+```{r echo=TRUE, eval=FALSE}
+all_values <- c(sample_values, outliers)
+z_scores <- (all_values - mean(all_values)) / sd(all_values)
+ggplot(data.frame(scaled = z_scores), aes(x = scaled)) + geom_histogram(binwidth = 0.2)
+```
+
+![](images/skewed.png){#id .class width=40% height=40%}
+
+## RAPIDS
+
+Readers following Apache Spark releases closely probably have noticed the recent addition of [RAPIDS](https://rapids.ai/) GPU acceleration support in Spark 3.0. Catching up with this recent development, an option to enable RAPIDS in Spark connections was also created in `sparklyr` and shipped in `sparklyr` 1.4. On a host with RAPIDS-capable hardware (e.g., an Amazon EC2 instance of type 'p3.2xlarge'), one can install `sparklyr` 1.4 and observe RAPIDS hardware acceleration being reflected in Spark SQL physical query plans:
+
+```{r echo=TRUE, eval=FALSE}
+library(sparklyr)
+
+sc <- spark_connect(master = "local", version = "3.0.0", packages = "rapids")
+dplyr::db_explain(sc, "SELECT 4")
+```
+
+```
+== Physical Plan ==
+*(2) GpuColumnarToRow false
++- GpuProject [4 AS 4#45]
+   +- GpuRowToColumnar TargetSize(2147483647)
+      +- *(1) Scan OneRowRelation[]
+```
+
+## Higher-Order Functions and `dplyr`-Related Improvements
+
+All newly introduced higher-order functions from Spark 3.0, such as `array_sort()` with custom comparator, `transform_keys()`, `transform_values()`, and `map_zip_with()`, are supported by `sparklyr` 1.4.
+
+In addition, all higher-order functions can now be accessed directly through `dplyr` rather than their `hof_*` counterparts in `sparklyr`, which means, for example, we can run the following `dplyr` queries to calculate the square of all array elements in column `x` of `sdf` and then sort them in descending order:
```

Comment:
```suggestion
In addition, all higher-order functions can now be accessed directly through `dplyr` rather than their `hof_*` counterparts in `sparklyr`. This means, for example, that we can run the following `dplyr` queries to calculate the square of all array elements in column `x` of `sdf`, and then sort them in descending order:
```